File: Namyata/LastStoneWeight.py
import unittest
from typing import List
from bisect import insort_left
# You are given an array of integers stones where stones[i] is the weight of the ith stone.
#
# We are playing a game with the stones. On each turn, we choose the heaviest two stones and smash them together. Suppose the heaviest two stones have weights x and y with x <= y. The result of this smash is:
#
# If x == y, both stones are destroyed, and
# If x != y, the stone of weight x is destroyed, and the stone of weight y has new weight y - x.
# At the end of the game, there is at most one stone left.
#
# Return the weight of the last remaining stone. If there are no stones left, return 0.
#
# Constraints:
#
# 1 <= stones.length <= 30
# 1 <= stones[i] <= 1000



def lastStoneWeight(stones: List[int]) -> int:
    stones.sort()

    while stones:
        s1 = stones.pop()

        if not stones:
            return s1

        s2 = stones.pop()

        if s1 > s2:
            insort_left(stones, s1 - s2)

    return 0

class TestLastStoneWeight(unittest.TestCase):
    def test_single_stone(self):
        stones = [5]
        expected = 5
        self.assertEqual(lastStoneWeight(stones), expected)

    def test_two_stones(self):
        stones = [3, 7]
        expected = 4
        self.assertEqual(lastStoneWeight(stones), expected)

    def test_multiple_stones(self):
        stones = [2, 7, 4, 1, 8, 1]
        expected = 1
        self.assertEqual(lastStoneWeight(stones), expected)

    def test_same_weight_stones(self):
        stones = [5, 5, 5, 5]
        expected = 0
        self.assertEqual(lastStoneWeight(stones), expected)

    def test_empty_list(self):
        stones = []
        expected = 0
        self.assertEqual(lastStoneWeight(stones), expected)

if __name__ == '__main__':
    unittest.main()
# all the test cases passed.

File: CTests/last_stone_weight.c
#include <stdio.h>
#include <stdbool.h>
#include <assert.h>
#include <stdlib.h>

int compareInt(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int lastStoneWeight(int stones[], int n) {
    qsort(stones, n, sizeof(int), compareInt);

    while (n > 1) {
        int s1 = stones[n - 1];
        int s2 = stones[n - 2];
        n -= 2;

        if (s1 > s2) {
            stones[n] = s1 - s2;
            n++;
            qsort(stones, n, sizeof(int), compareInt);
        }
    }

    return (n == 1) ? stones[0] : 0;
}

void test_single_stone() {
    int stones[] = {5};
    int n = sizeof(stones) / sizeof(stones[0]);
    int result = lastStoneWeight(stones, n);
    assert(result == 5);
    printf("Test case 'single_stone' passed\n");
}

void test_two_stones() {
    int stones[] = {3, 7};
    int n = sizeof(stones) / sizeof(stones[0]);
    int result = lastStoneWeight(stones, n);
    assert(result == 4);
    printf("Test case 'two_stones' passed\n");
}

void test_multiple_stones() {
    int stones[] = {2, 7, 4, 1, 8, 1};
    int n = sizeof(stones) / sizeof(stones[0]);
    int result = lastStoneWeight(stones, n);
    assert(result == 1);
    printf("Test case 'multiple_stones' passed\n");
}

void test_same_weight_stones() {
    int stones[] = {5, 5, 5, 5};
    int n = sizeof(stones) / sizeof(stones[0]);
    int result = lastStoneWeight(stones, n);
    assert(result == 0);
    printf("Test case 'same_weight_stones' passed\n");
}

void test_empty_list() {
    int stones[] = {};
    int n = sizeof(stones) / sizeof(stones[0]);
    int result = lastStoneWeight(stones, n);
    assert(result == 0);
    printf("Test case 'empty_list' passed\n");
}

int main() {
    test_single_stone();
    test_two_stones();
    test_multiple_stones();
    test_same_weight_stones();
    test_empty_list();
    return 0;
}
